src/uct/ib/dc/dc_mlx5.c

```diff
@@ -773,67 +779,137 @@ static void uct_dc_mlx5_iface_dcis_destroy(uct_dc_mlx5_iface_t *iface,
                                     iface->tx.dcis[dci_index].txwq.reg);
     }
 
+    iface->tx.dci_counter = 0;
+
     for (pool_index = 0; pool_index < num_dci_pools; pool_index++) {
         uct_dc_mlx5_iface_dci_pool_destroy(&iface->tx.dci_pool[pool_index]);
     }
 
     ucs_free(iface->tx.dcis);
 }
 
+ucs_status_t
+uct_dc_mlx5_iface_create_dci_pool(uct_dc_mlx5_iface_t *iface,
+                                  const uct_dc_mlx5_dci_config_t *config,
+                                  uint8_t pool_size, uint8_t *pool_index_p)
+{
+    uint8_t full_handshake = iface->flags &
+                             UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE_FULL_HANDSHAKE;
+    uint8_t num_dcis       = iface->tx.ndci;
+    int pool_index         = iface->tx.num_dci_pools;
+
+    uct_dc_mlx5_dci_pool_t *dci_pool;
+    ucs_status_t status;
+    int i;
+
+    ucs_assert(iface->tx.num_dci_pools < UCT_DC_MLX5_IFACE_MAX_DCI_POOLS);
+    ucs_debug("creating dci pool %u with %u QPs", pool_index, num_dcis);
+    dci_pool        = &iface->tx.dci_pool[pool_index];
+    dci_pool->stack = ucs_calloc(pool_size, sizeof(*dci_pool->stack),
+                                 "dci pool stack");
+    if (dci_pool->stack == NULL) {
+        ucs_error("failed to allocate dci_pool stack");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ucs_arbiter_init(&dci_pool->arbiter);
+    dci_pool->stack_top         = 0;
+    dci_pool->release_stack_top = -1;
+
+    for (i = 0; i < pool_size; ++i) {
+        status = uct_dc_mlx5_iface_create_dci(
+                iface, pool_index, iface->tx.dci_counter,
+                config->key.path_index % iface->super.super.super.num_paths,
+                full_handshake);
+        if (status != UCS_OK) {
+            ucs_error("failed to create dci: %d, dci_counter: %u", i,
+                      iface->tx.dci_counter);
+            goto err_pool_destroy;
+        }
+        dci_pool->stack[i] = iface->tx.dci_counter - 1;
+    }
+
+    iface->tx.num_dci_pools++;
+    *pool_index_p = pool_index;
+
+    return UCS_OK;
+
+err_pool_destroy:
+    uct_dc_mlx5_iface_dci_pool_destroy(dci_pool);
+    uct_dc_mlx5_iface_dcis_destroy(iface, pool_index, iface->tx.dci_counter);
+
+    return status;
+}
+
+/**
+ * Checks wether dci pool config present in the dc_config_hash and then either returning
```

Comment:
Checks whether dci pool config is present in dc_config_hash and returns the matching pool index or creates a new one

Reply:
Move docs to `.h`